src/app/abstracts/game.abstract.ts

```diff
@@ -0,0 +1,54 @@
+import { User } from 'discord.js';
+import { IContainer } from '../../common/types';
+import { GameResult, GameType } from '../../services/gameleaderboard.service';
+export default abstract class Game {
+  private _container: IContainer;
+  private _game: GameType;
+
+  constructor(container: IContainer, game: GameType) {
```

Comment:
container is normally public in here, that way you dont have to set it to a private field

Reply:
How would we give the container to the game?

Reply:
This is what it normally looks like
```
constructor(public container: IContainer) {
  super();
}
```
And you can just access it with this.container

Reply:
What would the declaration look like? If container wasn't declared, then it'll just say that's an unused variable.

Reply:
You mean in your case? Because the code I put is what we've always been using

Reply:
Yes in this case.

Reply:
I guess a better way to ask the question is.

Do the variables live in the abstract? If so, how do we get them there without passing them through super().

Reply:
I'm need to brush up on inheritance. It's been a while since I really had to do this stuff.
Is there and issue with passing it in super?

Reply:
> I guess a better way to ask the question is.
> 
> Do the variables live in the abstract? If so, how do we get them there without passing them through super().

Yeah, if you initialize/instantiate a variable in an abstract class that value is carried down to extending classes. So in the TicTacToe plugin you can simply use `this.container`

Reply:
Awesome this is how I thought things worked! This is how it is currently currently implemented, with necessary data living in the Abstract class, and being passed in through the constructor.

So the question still lies though, @tanndlin  what changes are you suggesting here.

Reply:
I think he wants your container field to be either public or protected, as it stands right now, subclasses can't access the container field.

Reply:
Ah I see, I currently only have playerA and playerB as protected members, and keep everything else as private.

What is the benefit of publicizing the container? Isn't the whole point of the abstraction to keep game logic in one place and system logic in another?

Reply:
In this case we aren't really concerned with encapsulation when it comes to inheritance. We are only concerned about it when it's used externally. When I say "externally" I mean as value within your code, not by extending it. Your code is encapsulated from the outside. But that same encapsulation doesn't apply for inheritance. Extending classes still have to override the abstract methods. And they have access to the protected fields. Contrast this with using an instance of `Game` externally, all of the protected fields are successfully hidden.

This is more of a template than an encapsulation. Almost every property in `Game` can be overwritten anyways just by extending it. This is fine because extending classes sometimes need that flexibility depending on their use case. But if they don't need that flexibility the `Game` class offers reasonable defaults (e.g. the non-abstract methods).

Reply:
Ah understood! So along that same vein, we probably should keep Game variables protected rather than fully public, as to discourage changes being made from outside of a Game instance.

What are your thoughts?

Reply:
Sounds good to me!